cli/src/complete.rs

```diff
@@ -446,6 +448,141 @@ pub fn leaf_config_keys() -> Vec<CompletionCandidate> {
     config_keys_impl(true)
 }
 
+fn all_files_from_rev(rev: String) -> Vec<CompletionCandidate> {
+    with_jj(|jj, _| {
+        let mut child = jj
+            .build()
+            .arg("file")
+            .arg("list")
+            .arg("--revision")
+            .arg(rev)
+            .stdout(std::process::Stdio::piped())
+            .spawn()
+            .map_err(user_error)?;
+        let stdout = child.stdout.take().unwrap();
+
+        Ok(std::io::BufReader::new(stdout)
+            .lines()
+            .take(1_000)
+            .map_while(Result::ok)
+            .map(CompletionCandidate::new)
+            .collect())
+    })
+}
+
+fn modified_files_from_rev_with_jj_cmd(
+    rev: (String, Option<String>),
+    mut cmd: std::process::Command,
+) -> Result<Vec<CompletionCandidate>, CommandError> {
+    cmd.arg("diff").arg("--summary");
+    match rev {
+        (rev, None) => cmd.arg("--revision").arg(rev),
+        (from, Some(to)) => cmd.arg("--from").arg(from).arg("--to").arg(to),
+    };
+    let output = cmd.output().map_err(user_error)?;
+    let stdout = String::from_utf8_lossy(&output.stdout);
+
+    Ok(stdout
+        .lines()
+        .map(|line| {
+            let (mode, path) = line
+                .split_once(' ')
+                .expect("diff --summary should contain a space between mode and path");
+            let help = match mode {
+                "M" => "Modified".into(),
+                "D" => "Deleted".into(),
+                "A" => "Added".into(),
+                "R" => "Renamed".into(),
+                "C" => "Copied".into(),
+                _ => format!("unknown mode: '{mode}'"),
+            };
+            CompletionCandidate::new(path).help(Some(help.into()))
+        })
+        .collect())
+}
+
+fn modified_files_from_rev(rev: (String, Option<String>)) -> Vec<CompletionCandidate> {
+    with_jj(|jj, _| modified_files_from_rev_with_jj_cmd(rev, jj.build()))
+}
+
+fn conflicted_files_from_rev(rev: &str) -> Vec<CompletionCandidate> {
+    with_jj(|jj, _| {
+        let output = jj
+            .build()
+            .arg("resolve")
+            .arg("--list")
+            .arg("--revision")
+            .arg(rev)
+            .output()
+            .map_err(user_error)?;
+        let stdout = String::from_utf8_lossy(&output.stdout);
+
+        Ok(stdout
+            .lines()
+            .filter_map(|line| line.split_whitespace().next())
+            .map(CompletionCandidate::new)
+            .collect())
+    })
+}
+
+pub fn modified_files() -> Vec<CompletionCandidate> {
+    modified_files_from_rev(("@".into(), None))
+}
+
+pub fn all_revision_files(current: &std::ffi::OsStr) -> Vec<CompletionCandidate> {
+    if current.is_empty() {
+        // Avoid too much work in big repositories.
```

Comment:
nit: perhaps, this comment no longer makes sense? We do the same amount of work no matter if `current` is empty or not, and the output is capped at 1000 lines.

(we shouldn't remove the `current` argument btw, we'll need to pass it as a glob pattern or something if `jj file list` gains an option to list directory paths.)